bpf/lib/strict.h

```diff
@@ -0,0 +1,68 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Cilium */
+
+#ifndef __STRICT_H_
+#define __STRICT_H_
+
+#ifdef ENCRYPTION_STRICT_MODE
+
+#include <bpf/ctx/ctx.h>
+#include <bpf/api.h>
+#include <linux/tcp.h>
+#include <linux/ip.h>
+#include <linux/ipv6.h>
+
+#include "common.h"
+#include "dbg.h"
+#include "overloadable.h"
+
+/* strict_allow checks whether the packet is allowed to pass through the strict mode. */
+static __always_inline bool
+strict_allow(struct __ctx_buff *ctx) {
+	struct remote_endpoint_info __maybe_unused *dest_info, __maybe_unused *src_info;
+	bool __maybe_unused in_strict_cidr = false;
+	void *data, *data_end;
+#ifdef ENABLE_IPV4
+	struct iphdr *ip4;
+#endif
+	__u16 proto = 0;
+
+	if (!validate_ethertype(ctx, &proto))
+		return true;
+
+	switch (proto) {
+#ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return true;
+
+		/* Allow traffic that is send from the node:
+		 * (1) When encapsulation is used and the destination is a remote pod.
+		 * (2) When the destination is a remote-node.
+		 */
+		if (ip4->saddr == IPV4_GATEWAY || ip4->saddr == IPV4_ENCRYPT_IFACE)
+			return true;
+
+		in_strict_cidr = ipv4_is_in_subnet(ip4->daddr,
+						   STRICT_IPV4_NET,
+						   STRICT_IPV4_NET_SIZE);
+		in_strict_cidr &= ipv4_is_in_subnet(ip4->saddr,
+						    STRICT_IPV4_NET,
+						    STRICT_IPV4_NET_SIZE);
+
+#if defined(TUNNEL_MODE) || defined(STRICT_IPV4_OVERLAPPING_CIDR)
+		/* Allow pod to remote-node communication */
+		dest_info = lookup_ip4_remote_endpoint(ip4->daddr, 0);
+		if (dest_info && dest_info->sec_label &&
+		    identity_is_node(dest_info->sec_label))
+			return true;
+#endif /* TUNNEL_MODE || STRICT_IPV4_OVERLAPPING_CIDR */
+		return !in_strict_cidr;
+#endif /* ENABLE_IPV4 */
+	default:
+		return true;
```

Comment:
I almost overlooked the documentation that only IPv4 is supported. I think we should mention in the docs that IPv6 traffic can be leaked.

I would also consider adding a warning in the agent if IPv6 is enabled together with strict mode, since I think it's rate that users want to lock down IPv4 traffic, but don't care about IPv6? Unless I'm missing something that that is a very common usecase.

Reply:
Added another sentence about the implication that the strict mode only supports IPv4, namely that it's encryption is not enforced and also the warning in the agent.

We already had a quick look at IPv6 support in an earlier version. This is probably the first feature I'd add after this PR has been merged. Since this is also needed to eventually move the strict mode integration tests into the `cilium-cli` repository, as at least from what I saw, there both are checked by default. 

From a technical point of view, the idea is to first support 2 CIDRs for dual IPv4,6 setups. Eventually we could support a list of strict mode CIDRs of mixed IPv4 and IPv6 of custom length. But then we'd have to use a bpf map to hand over the data, I think.